homeassistant/components/purpleair/__init__.py

```diff
@@ -23,9 +25,35 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
+    entry.async_on_unload(entry.add_update_listener(async_handle_entry_update))
+
     return True
 
 
+async def async_handle_entry_update(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle an options update."""
+    if entry.options.get(CONF_LAST_UPDATE_SENSOR_ADD) is True:
```

Comment:
Do we really need this check? Can't we always reload when the options are updated? It probably won't happen frequently.

Reply:
Without this check, deleting an entry (and subsequently attempting to reload the config entry) logs this error:

```
2022-12-20 09:54:53.842 ERROR (MainThread) [homeassistant.config_entries] Error unloading entry 2C468 for sensor
Traceback (most recent call last):
  File "/Users/bachya/Git/home-assistant/homeassistant/config_entries.py", line 515, in async_unload
    result = await component.async_unload_entry(hass, self)
  File "/Users/bachya/Git/home-assistant/homeassistant/components/sensor/__init__.py", line 480, in async_unload_entry
    return await component.async_unload_entry(entry)
  File "/Users/bachya/Git/home-assistant/homeassistant/helpers/entity_component.py", line 183, in async_unload_entry
    await platform.async_reset()
  File "/Users/bachya/Git/home-assistant/homeassistant/helpers/entity_platform.py", line 707, in async_reset
    await asyncio.gather(*tasks)
  File "/Users/bachya/Git/home-assistant/homeassistant/helpers/entity.py", line 813, in async_remove
    raise HomeAssistantError(
homeassistant.exceptions.HomeAssistantError: Entity sensor.sensor_name_humidity async_remove called twice
```

Reply:
Adding this option feels like a workaround to solve the race. It would be better if we can stop the race from occurring in the options flow that is the source.

Reply:
Open to suggestions on how to do that. When I originally encountered this, I didn't see any other options flows that deleted devices and then reloaded.

Reply:
Maybe we need a device entry registry helper that can await the removal of a device with all connected entities?

Reply:
That sounds much better than my idea of `await asyncio.sleep(1)`. 😂

Reply:
We've discussed this in the NabuCasa core meeting and said that we don't want to add a helper until it's a more common problem.

Another way how we could solve this in the integration could be to listen for the entity state change event in the config flow. Do the listen after we remove the device and not exit the flow until the entity state is `None`, ie the entity is removed. An `asyncio.Event` could be used to await the state change callback.

Reply:
https://github.com/home-assistant/core/pull/85086